### PR TITLE
feat(v2): support syncing tab choices

### PR DESCRIPTION
**Pull request by SamChou19815**
## Motivation

Implement feature requested in #2349, but in a more general way.

The linked issue wants auto-syncing tabs for programming languages. However, I think it can be implemented in a more general way. For example, we might also use the tabs to separate OS-specific instructions. Therefore, instead of providing hooks like `useProgrammingLanguages`, I decide to provide a more general hook called `useTabGroupChoiceContext`.

No matter whether we decide to limit the feature to programming languages or general tabs, we need to worry about conflicts. For example, inside the same project, the docs might have both tabs for JavaScript vs TypeScript, and tabs for NPM vs Yarn. We don't want these two different tabs to interfere with each other in surprising ways.

To prevent unwanted interference, I decide to separate these choices by keys. Users must explicit provide the same groupId to convey to docusaurus that those tabs are related and their choices should be synced. To preserve backward compatibility, groupId is an optional prop of Tab. Not passing the prop means that the user opts out of tab choice syncing.

The choice is stored persistently in `localStorage`, just like the choice of theme.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work. Bonus points for screenshots and videos!)

I added more examples of using tabs with `groupId` in `website/docs/markdown-features.mdx` to test the new feature.

example
example 2

(Some examples are silly, and I'm open to better examples.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/docusaurus, and link to your PR here.)

### CODE REVIEW

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.js [hunk 1]
```diff
@@ -17,8 +18,27 @@ const keys = {
 };
 
 function Tabs(props) {
-  const {block, children, defaultValue, values} = props;
+  const {block, children, defaultValue, values, groupId} = props;
+  const {getTabGroupChoice, setTabGroupChoice} = useTabGroupChoiceContext();
   const [selectedValue, setSelectedValue] = useState(defaultValue);
+
+  if (typeof groupId !== 'undefined') {
```
Comment (lex111) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.js]: It seems to me that we can remove the `groupId` prop (this is somewhat redundant) in favor of `value` prop in `TabItem` component to identify each tab? Yes, this way we synchronize tabs by default and do not give users prevent this behavior. (Although I do not think that is a big problem).

File: packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+import {useState, useCallback, useEffect} from 'react';
+
+const useTabGroupChoice = () => {
+  const [choices, setChoices] = useState({});
+  const setChoiceSyncWithLocalStorage = useCallback(
+    newChoices => {
+      try {
+        localStorage.setItem('tab-group', JSON.stringify(newChoices));
```
Comment (lex111) [inline packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: First, put storage value in separate const outside the hook-level. Second, add a namespace to that, for example `docusaurus.tab-group`.

File: packages/docusaurus-theme-classic/src/theme/TabGroupChoiceContext.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React from 'react';
+
+const TabGroupChoiceContext = React.createContext({
```
Comment (lex111) [inline packages/docusaurus-theme-classic/src/theme/TabGroupChoiceContext.js]: Maybe use this approach?

```js
import React, { createContext } from 'react';

const TabGroupChoiceContext = createContext(...)
```

**Comment by SamChou19815** [comment]
Updated the PR is store only one choice

File: packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js (hunk 2 above)
Comment (lex111) [reply packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: > put storage value in separate const outside the hook-level.

:up: @SamChou19815

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.js [hunk 4]
```diff
@@ -18,7 +19,24 @@ const keys = {
 
 function Tabs(props) {
   const {block, children, defaultValue, values} = props;
+  const {tabGroupChoice, setTabGroupChoice} = useTabGroupChoiceContext();
   const [selectedValue, setSelectedValue] = useState(defaultValue);
+
+  console.log({values, tabGroupChoice});
```
Comment (lex111) [inline packages/docusaurus-theme-classic/src/theme/Tabs/index.js]: Logging should be removed

File: packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js (hunk 2 above)
Comment (SamChou19815) [reply packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: @lex111 Do you mean moving the literal`'docusaurus.tab-choice'` outside of the hooks?

File: packages/docusaurus-theme-classic/src/theme/Tabs/index.js (hunk 4 above)
Comment (SamChou19815) [reply packages/docusaurus-theme-classic/src/theme/Tabs/index.js]: Oops. I will remove it.

File: packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js (hunk 2 above)
Comment (lex111) [reply packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: Something like this (come up with a better name for const):

```js
const STORAGE_KEY = ...

const setChoiceSyncWithLocalStorage = ...
```

File: packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import {useState, useCallback, useEffect} from 'react';
+
+const TAB_CHOICE_STORAGE_KEY = 'docusaurus.tab-choice';
+
+const useTabGroupChoice = () => {
+  const [tabGroupChoice, setChoice] = useState();
+  const setChoiceSyncWithLocalStorage = useCallback(
+    newChoice => {
+      try {
+        localStorage.setItem(TAB_CHOICE_STORAGE_KEY, newChoice);
```
Comment (lex111) [inline packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: `tab-choice` is must be an array in order to maintain sync for several different tabs. 

See https://deploy-preview-2366--docusaurus-2.netlify.com/docs/markdown-features/ for example.

image


image

Comment (SamChou19815) [reply packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: I think a plain array can work if all the tab choices happen to be disjoint. However, if we have something like `['win', 'mac']` and `['win', 'unix']` and we see `['win']` in the array, there is an ambiguity about which `win` it is referring to.

Should we make it an object keyed by serialized `values` array?

Comment (lex111) [reply packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: Not sure if this is necessary, so regular array should work. We assume that users will use the same tab names for similar tabs (Yarn VS npm eg) throughout the website.

File: packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import {useState, useCallback, useEffect} from 'react';
+
+const TAB_CHOICE_STORAGE_KEY = 'docusaurus.tab-choice';
```
Comment (yangshun) [inline packages/docusaurus-theme-classic/src/theme/hooks/useTabGroupChoice.js]: `docusaurus.tabs` would be sufficient.

**Pull request merged**